docs/RFCS/20180411_alter_column_type.md

```diff
@@ -0,0 +1,429 @@
+* The shadow column and indexes are upgraded to `WRITE_ONLY` mode.
+* Nodes that write to the source column will synchronously apply the
+  requested type change and duplicate the write to the shadow column and
+  indexes.
+* A backfill process is distributed across all nodes in the cluster
+  which have data for the table. This backfill will populate the shadow
+  column and shadow indexes, inserting data with the source MVCC
+  timestamp. The backfilled sstables will be inserted into the LSM such
+  that any shadowed write will have priority.
+* The following steps will only occur if the change is in auto-complete
+  mode or when the user issues the `COMPLETE` command.
+* Once the backfill is complete, the `WRITE_ONLY` restriction is removed
+  from the shadow column and shadow indexes. Any code that resolves a
+  column or index name to an id should be prepared to look for a shadow
+  that has been activated.
+* The table metadata is updated, changing the column name and indexes to
+  point at the shadow data, making the original column or indexes
+  non-live.
+* The old column and indexes are retired in the same manner as dropping
+  a column, allowing any active readers to complete their transaction.
+
+To summarize:
+* Column and index writers should be prepared to duplicate their writes
+  when a writable shadow exists for their target column or index.
+* Code that resolves a column or index name for reading should be
+  prepared to look for an activated shadow.
+* It must be possible to insert bulk, backfilled data into the LSM such
+  that the bulk data will be shadowed by any ad-hoc writes that occurred
+  during the backfilling process.
+
+## Failure modes
+
+Changing a column type is a low-risk activity, at least from the
+perspective of data safety, because the process can be aborted at any
+time up until the shadow column and indexes are promoted to `READ` mode.
+All validation or conversion of existing will have been performed before
+the new type becomes visible to SQL consumers, while any writes
+performed during the conversion process must conform to both the old and
+new column types.
+
+The most likely cause for type-change failure is the presence of
+existing data which cannot be coerced into the desired type. In this
+case, the type change should abort and a sample of non-conforming rows /
+values should be returned to the user.
+
+Type changes which require extensive rewriting of data will temporarily
+increase the amount of disk space required for the new sstables until
+the old data can be retired. The amount of extra disk space required
+could be estimated from existing table stats.
+
+In the case where a column-type conversion (e.g. `STRING -> UUID`) is in
+process and a user attempts to insert data that is not compatible with
+the target type, the `INSERT` should fail with an error message such as
+`column user_id is being converted from STRING to UUID. The value
+'foobar' cannot be converted to UUID.`
+
```

Comment:
The biggest complexity in this change is introducing the notion of shadow column everywhere in the run time code. Its quite similar to a computed column with the same name, I'd like us try to hide the name from the shadow column until it has been promoted, so that we don't have to make the assumption everywhere in the code that a duplicate column can exist. perhaps we should leave its name unspecified until the promotion happens.

Reply:
@justinj might have some suggestions here.